### PR TITLE
(Bugfix) Encode url for albums

### LINKED ISSUE #89: Wrong URL link when using special characters in folder title
When naming an album with a question mark for instance "Question ?", the generated url is wrong and the user is redirected to an empty folder.

image

The "%3F" is missing in the generated url:
image

@skjnldsv is the automatic redirection when visiting an empty album intended?

### PR DESCRIPTION
**Pull request by Mikescops**
Try to fix #89. 
I try to encode the url at the folder level, but maybe we need to make this earlier in the "pipeline".

### CODE REVIEW

**Comment by Mikescops** [comment]
@skjnldsv I think there is a bigger issue here. We shouldn't have the url path using a user folder name. 

Here is what the file app do:
`apps/files/?dir=/Documents/question %3F&fileid=239`

Here is what the photo app do:
`apps/photos/albums/Documents/question ?`

2 issues:
1. The path is not encoded that can lead to injections in url 
1. Also the question mark `?` is understood as the query parameters separator

**Comment by skjnldsv** [comment]
> * The path is not encoded that can lead to injections in url

We do not handle the path other than as a string, so I don't think you can inject anything with it. Worse case, you end up with a wrongly formatted url and the server files listing request will fail. Were you able to generate something?


> 2\. Also the question mark `?` is understood as the query parameters separator

I'm fine with encoding everything aside slashes :)
I feel like it's nice to have a visually pleasant url

What do you think? Wan we work around this?

**Comment by Mikescops** [comment]
Problem is that the question mark encoded or not will be understood by the browser as the start of query parameters.
So when providing `"apps/photos/albums/Documents/question ?"`, nextcloud app will get `Documents/question%20` for the path.

**Comment by Mikescops** [comment]
For me we are forced to do it that way `?dir=/Documents/question %3F` using a query parameter

**Comment by skjnldsv** [comment]
> So when providing `"apps/photos/albums/Documents/question ?"`, nextcloud app will get `Documents/question%20` for the path.

looks good to me?

**Comment by Mikescops** [comment]
> > So when providing `"apps/photos/albums/Documents/question ?"`, nextcloud app will get `Documents/question%20` for the path.
> 
> looks good to me?

Nope we are missing the question mark, then nextcloud cannot find the folder. Am i clear?

**Comment by skjnldsv** [comment]
> Nope we are missing the question mark, then nextcloud cannot find the folder. Am i clear?

Right, I misread your comment.
What about https://dev.test.com/apps/photos/Test%20%3F ?

**Comment by Mikescops** [comment]
Not working the browser interpret the %3F as ? on the fly

**Comment by skjnldsv** [comment]
> Not working the browser interpret the %3F as ? on the fly

For me it's working :)
Capture d’écran_2020-01-07_17-13-20

**Comment by skjnldsv** [comment]
But if it's too much work, let's switch back to using ?path=

**Comment by Mikescops** [comment]
When you direct access or when you click on a href link?

**Comment by skjnldsv** [comment]
> When you direct access or when you click on a href link?

direct

**Comment by Mikescops** [comment]
Yep when using a <a> tag the browser unencode %3F

**Comment by Mikescops** [comment]
Ok I found a fix, that's the best i can come with, you may want to find a better solution tho'

**Comment by Mikescops** [comment]
Maybe we could unit test this :+1:

**Comment by skjnldsv** [comment]
> Maybe we could unit test this

I dream of having the time to do this :p

File: src/utils/fileUtils.js [hunk 1]
```diff
@@ -118,6 +118,10 @@ const genFileInfo = function(obj) {
 					: data
 			}
 		}
+
+		if (key === 'filename') {
+			fileInfo[camelcase(key)] = escape(data)
```
Comment (skjnldsv) [inline src/utils/fileUtils.js]: We should avoid touching the fileinfos. They should stay as pristine as what the dav backend is giving us :)
I would rather have this fixed in  the vue router side: https://github.com/nextcloud/photos/blob/7ed958ced085f4548dafe7646d26e801e39732e4/src/router/index.js#L57

Comment (Mikescops) [reply src/utils/fileUtils.js]: I'll try

**Comment by rullzer** [comment]
Can't we do it the stupid approach..

`..../FILEID/PATH`.

Then the code can use the fileid as the actual location. But still have the path as fancy pancy that you see where you are...

File: src/utils/fileUtils.js (hunk 1 above)
Comment (Mikescops) [reply src/utils/fileUtils.js]: Can't find a way to do this..

Comment (skjnldsv) [reply src/utils/fileUtils.js]: Let's talk tomorrow :)
You're here already?

Comment (Mikescops) [reply src/utils/fileUtils.js]: I'll be there on Friday :/

Comment (skjnldsv) [reply src/utils/fileUtils.js]: Oh, you're only there for two days?

Comment (Mikescops) [reply src/utils/fileUtils.js]: Yes, I'll explain you why, but we're underwater at work

Comment (skjnldsv) [reply src/utils/fileUtils.js]: Arf :(
I'll figure it out :)

**Pull request closed without merging**